src/EditorFeatures/Core/Implementation/NavigationBar/NavigationBarController.cs

```diff
@@ -70,80 +73,58 @@ internal partial class NavigationBarController : ForegroundThreadAffinitizedObje
 
             _modelTask = Task.FromResult(_latestModelAndSelectedInfo_OnlyAccessOnUIThread.model);
+
+            _eventSource = new CompilationAvailableTaggerEventSource(
+                subjectBuffer,
+                asyncListener,
+                TaggerEventSources.OnTextChanged(subjectBuffer),
+                TaggerEventSources.OnDocumentActiveContextChanged(subjectBuffer),
+                TaggerEventSources.OnWorkspaceChanged(subjectBuffer, asyncListener),
+                TaggerEventSources.OnWorkspaceRegistrationChanged(subjectBuffer));
+            _eventSource.Changed += OnEventSourceChanged;
+            _eventSource.Connect();
         }
 
         public void SetWorkspace(Workspace? newWorkspace)
         {
-            DisconnectFromWorkspace();
-
             if (newWorkspace != null)
-            {
-                ConnectToWorkspace(newWorkspace);
-            }
+                StartModelUpdateAndSelectedItemUpdateTasks();
         }
 
-        private void ConnectToWorkspace(Workspace workspace)
+        private void StartModelUpdateAndSelectedItemUpdateTasks()
         {
-            // If we disconnected before the workspace ever connected, just disregard
+            // If we're disconnected , just disregard
             if (_disconnected)
-            {
                 return;
-            }
-
-            _workspace = workspace;
-            _workspace.WorkspaceChanged += this.OnWorkspaceChanged;
-            _workspace.DocumentActiveContextChanged += this.OnDocumentActiveContextChanged;
 
             if (IsForeground())
             {
-                ConnectToNewWorkspace();
+                StartModelUpdateAndSelectedItemUpdateTasksOnUIThread();
             }
             else
             {
-                var asyncToken = _asyncListener.BeginAsyncOperation(nameof(ConnectToWorkspace));
+                var asyncToken = _asyncListener.BeginAsyncOperation(nameof(StartModelUpdateAndSelectedItemUpdateTasks));
                 Task.Run(async () =>
                 {
                     await ThreadingContext.JoinableTaskFactory.SwitchToMainThreadAsync();
-
-                    ConnectToNewWorkspace();
+                    StartModelUpdateAndSelectedItemUpdateTasksOnUIThread();
                 }).CompletesAsyncOperation(asyncToken);
             }
-
-            return;
-
-            void ConnectToNewWorkspace()
-            {
-                // For the first time you open the file, we'll start immediately
-                StartModelUpdateAndSelectedItemUpdateTasks(modelUpdateDelay: 0);
-            }
         }
 
-        private void DisconnectFromWorkspace()
+        private void OnEventSourceChanged(object? sender, TaggerEventArgs e)
         {
-            if (_workspace != null)
-            {
-                _workspace.DocumentActiveContextChanged -= this.OnDocumentActiveContextChanged;
-                _workspace.WorkspaceChanged -= this.OnWorkspaceChanged;
```

Comment:
much simpler.  no more explicit event hookups on our end to a lot of disparate sources.